crates/provider/src/chain.rs

```diff
@@ -106,8 +112,77 @@ impl<T: Transport + Clone, N: Network> ChainStreamPoller<T, N> {
                     }
                 };
                 self.known_blocks.put(number, block);
+                if self.known_blocks.len() == BLOCK_CACHE_SIZE.get() {
+                    // Cache is full, should be consumed before filling more blocks.
+                    debug!(number, "cache full");
+                    break;
+                }
             }
         }
         }
     }
 }
+
+#[cfg(all(test, feature = "anvil-api"))] // Tests rely heavily on ability to mine blocks on demand.
```

Comment:
Writing tests without `anvil_mine` would be more complex, so I hope it's fine given that tests are run with `--all-features` in CI.